File: thesis/algorithms/go/np-dificil.go
package main

import (
    "encoding/json"
    "fmt"
    "io/ioutil"
    "os"
    "strings"
)

type Program struct {
    Program string `json:"program"`
}

func simulate(code string) bool {
    return strings.ToUpper(strings.TrimSpace(code)) == "HALT"
}

func main() {
    size := "small"
    if len(os.Args) > 1 {
        size = os.Args[1]
    }
    path := "../datasets/" + size + "/halting_" + size + ".json"

    data, err := ioutil.ReadFile(path)
    if err != nil {
        fmt.Println("Erro ao ler arquivo:", err)
        return
    }

    var programs []Program
    if err := json.Unmarshal(data, &programs); err != nil {
        fmt.Println("Erro ao decodificar JSON:", err)
        return
    }

    halted := 0
    for _, p := range programs {
        if simulate(p.Program) {
            halted++
        }
    }

    fmt.Printf("%d de %d programas halting (%s)\n", halted, len(programs), size)
}

File: thesis/algorithms/kotlin/np-dificil.kt
import java.io.File
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue

data class Program(val program: String)

fun simulate(code: String): Boolean {
    return code.trim().uppercase() == "HALT"
}

fun main(args: Array<String>) {
    val size = if (args.isNotEmpty()) args[0] else "small"
    val path = "datasets/$size/halting.json"
    val content = File(path).readText()
    val mapper = jacksonObjectMapper()
    val programs: List<Program> = mapper.readValue(content)
    val halted = programs.count { simulate(it.program) }
    println("$halted de ${programs.size} programas halting ($size)")
}
